File: lib/pagination.ts
const DEFAULT_PAGE = 1
const DEFAULT_PAGE_SIZE = 25

type PaginationQuery = {
  page?: number
  pageSize?: number
}

type Page<T> = {
  items: T[]
  page: number
  pageSize: number
  totalItems: number
  totalPages: number
}

type BuildPageInput<T> = {
  items: T[]
  page: number
  pageSize: number
  totalItems: number
}

export function resolvePagination(query?: PaginationQuery) {
  const page = query?.page ?? DEFAULT_PAGE
  const pageSize = query?.pageSize ?? DEFAULT_PAGE_SIZE

  return {
    page,
    pageSize,
    offset: (page - 1) * pageSize,
  }
}

export function buildPage<T>({ items, page, pageSize, totalItems }: BuildPageInput<T>): Page<T> {
  return {
    items,
    page,
    pageSize,
    totalItems,
    totalPages: totalItems === 0 ? 0 : Math.ceil(totalItems / pageSize),
  }
}

export { DEFAULT_PAGE, DEFAULT_PAGE_SIZE }
export type { Page, PaginationQuery }

File: backend/internal/pagination/pagination.go
package pagination

const DefaultPage = 1
const DefaultPageSize = 25

type Page[T any] struct {
	Items      []T `json:"items"`
	Page       int `json:"page"`
	PageSize   int `json:"pageSize"`
	TotalItems int `json:"totalItems"`
	TotalPages int `json:"totalPages"`
}

func Normalize(page, pageSize *int32) (int, int) {
	p := DefaultPage
	ps := DefaultPageSize
	if page != nil && *page > 0 {
		p = int(*page)
	}
	if pageSize != nil && *pageSize > 0 {
		ps = int(*pageSize)
	}
	return p, ps
}

func Offset(page, pageSize int) int {
	return (page - 1) * pageSize
}

func TotalPages(totalItems, pageSize int) int {
	if pageSize == 0 {
		return 0
	}
	total := totalItems / pageSize
	if totalItems%pageSize > 0 {
		total++
	}
	return total
}
